### PR TITLE
xds/ringhash: cache connectivity state of subchannels inside picker

### PR DESCRIPTION
**Pull request by dfawley**
I did not add a new test for this.  The sequence of operations required here is fairly long and specific and requires races that are too difficult to reliably stimulate.  Corresponding change in C++: [grpc/grpc@`98cdac3` (#32326)](https://github.com/grpc/grpc/pull/32326/commits/98cdac31a0196d23a972acd3de027a645daf66ce).  (gRPC-Java always kept a copy of the states in the picker.)

RELEASE NOTES:
* xds/ringhash: cache connectivity state of subchannels inside picker to avoid rare races

### CODE REVIEW

File: xds/internal/balancer/ringhash/ringhash.go [hunk 1]
```diff
@@ -347,37 +347,22 @@ func (b *ringhashBalancer) UpdateSubConnState(sc balancer.SubConn, state balance
...
-		b.regeneratePicker()
-		sendUpdate = true
 	case connectivity.TransientFailure:
 		// Save error to be reported via picker.
 		b.connErr = state.ConnectionError
-		b.regeneratePicker()
 	case connectivity.Shutdown:
 		// When an address was removed by resolver, b called RemoveSubConn but
 		// kept the sc's state in scStates. Remove state for this sc here.
 		delete(b.scStates, sc)
```
Comment (zasweq) [inline xds/internal/balancer/ringhash/ringhash.go:359]: Nit: Not a part of this PR, but this scStates field I feel like is wrongly named. This is a map from balancer.SubConn to the Ring Hashes SubConn with extra info. I feel like states is only a part of the reason this map exists.

File: xds/internal/balancer/ringhash/ringhash.go [hunk 2]
```diff
@@ -347,37 +347,22 @@ func (b *ringhashBalancer) UpdateSubConnState(sc balancer.SubConn, state balance
...
 	}
 
-	if sendUpdate {
+	if oldSCState != newSCState {
+		// Because the picker caches the state of the subconns, we always regnerate
```
Comment (zasweq) [inline xds/internal/balancer/ringhash/ringhash.go:363]: regenerate

File: xds/internal/balancer/ringhash/ringhash.go [hunk 3]
```diff
@@ -347,37 +347,22 @@ func (b *ringhashBalancer) UpdateSubConnState(sc balancer.SubConn, state balance
...
+		// and update the picker when the effective state changes.
```
Comment (zasweq) [inline xds/internal/balancer/ringhash/ringhash.go:364]: Nit: "of the SubConn." Optional addition: "This is so the Client Conn gets the most updated state as a former picker has a static cache of SubConn states at it's build time."

**Review by zasweq — APPROVED** [review]
LGTM.

File: xds/internal/balancer/ringhash/ringhash.go (hunk 2 above)
Comment (dfawley) [reply xds/internal/balancer/ringhash/ringhash.go]: Fixed

File: xds/internal/balancer/ringhash/ringhash.go (hunk 3 above)
Comment (dfawley) [reply xds/internal/balancer/ringhash/ringhash.go]: Fixed

File: xds/internal/balancer/ringhash/ringhash.go (hunk 1 above)
Comment (dfawley) [reply xds/internal/balancer/ringhash/ringhash.go:359]: Yeah it probably was changed during development.  99% of why there exists a map is state tracking.  It also does a `queueConnect` in rare cases.

**Pull request merged**